{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: EnsDb.Hsapiens.v86\n",
      "\n",
      "Loading required package: ensembldb\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: GenomicFeatures\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Possible Ensembl SSL connectivity problems detected.\n",
      "Please see the 'Connection Troubleshooting' section of the biomaRt vignette\n",
      "vignette('accessing_ensembl', package = 'biomaRt')Error in curl::curl_fetch_memory(url, handle = handle) : \n",
      "  SSL peer certificate or SSH remote key was not OK: [uswest.ensembl.org] SSL certificate problem: certificate has expired\n",
      "\n",
      "\n",
      "Loading required package: AnnotationFilter\n",
      "\n",
      "\n",
      "Attaching package: 'ensembldb'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "Loading required package: EnsDb.Mmusculus.v79\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: lattice\n",
      "\n",
      "Loading required package: SingleCellExperiment\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: 'matrixStats'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:Biobase':\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'MatrixGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:matrixStats':\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'dplyr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:matrixStats':\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:ensembldb':\n",
      "\n",
      "    filter, select\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationDbi':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:GenomicRanges':\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomeInfoDb':\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:IRanges':\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:S4Vectors':\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:BiocGenerics':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'gridExtra'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:dplyr':\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from 'package:BiocGenerics':\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'Matrix'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:S4Vectors':\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'rlang'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    exprs\n",
      "\n",
      "\n",
      "Attaching SeuratObject\n",
      "\n",
      "\n",
      "Attaching package: 'Seurat'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:SummarizedExperiment':\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'tidyr'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:Matrix':\n",
      "\n",
      "    expand, pack, unpack\n",
      "\n",
      "\n",
      "The following object is masked from 'package:S4Vectors':\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'xml2'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:rlang':\n",
      "\n",
      "    as_list\n",
      "\n",
      "\n",
      "Loading required package: viridisLite\n",
      "\n"
     ]
    }
   ],
   "source": [
    "source(\"~/sc-online/extraGenes.R\")\n",
    "source(\"~/sc-online/myExpSetCreatorFn.R\")\n",
    "source(\"~/sc-online/utils.R\")\n",
    "source(\"~/sc-online/labelTransfer.R\")\n",
    "\n",
    "library(caret)\n",
    "library(\"DropletUtils\")\n",
    "library(dplyr)\n",
    "library(ggplot2)\n",
    "library(grid)\n",
    "library(gridExtra)\n",
    "library(gtable)\n",
    "library(lisi)\n",
    "library(Matrix)\n",
    "library(patchwork)\n",
    "library(pheatmap)\n",
    "library(RColorBrewer)\n",
    "library(rhdf5)\n",
    "library(rlang)\n",
    "\n",
    "library(Seurat)\n",
    "library(tidyr)\n",
    "library(xml2)\n",
    "\n",
    "library(viridis)\n",
    "library(viridisLite)\n",
    "\n",
    "READ_RAW_READS = TRUE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reading donor_ids and sce files...\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reading Demuxlet donor assignments...\"\n"
     ]
    }
   ],
   "source": [
    "# load data from Vireo\n",
    "vireo_libs_dated = list(\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719B',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719BD',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719C',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719CD',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719D1',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719D2',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719DD',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719E1',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719E2',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719ED',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719F1',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719F2',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719FD',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719G1',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719G2',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719GD',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719I1',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719I2',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719ID',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719J1',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719J2',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719JD',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719K1',\n",
    "    '230712_pCalicoPDsHSrSNiPoold230719K2',\n",
    "    'pCalicoPDsHSrSNA8id230921A8',\n",
    "    'pCalicoPDsHSrSNA9id230921A9',\n",
    "    'pCalicoPDsHSrSNB8id230921B8',\n",
    "    'pCalicoPDsHSrSNB9id230921B9',\n",
    "    'pCalicoPDsHSrSNC8id230921C8',\n",
    "    'pCalicoPDsHSrSNC9id230921C9',\n",
    "    'pCalicoPDsHSrSND8id230921D8',\n",
    "    'pCalicoPDsHSrSND9id230921D9',\n",
    "    'pCalicoPDsHSrSNE8id230921E8',\n",
    "    'pCalicoPDsHSrSNF8id230921F8',\n",
    "    'pCalicoPDsHSrSNG8id230921G8',\n",
    "    'pCalicoPDsHSrSNH8id230921H8'\n",
    ")\n",
    "\n",
    "\n",
    "donor_ids_list=list()\n",
    "sce_list=list()\n",
    "\n",
    "base_path = \"/mnt/accessory/seq_data/calico\"\n",
    "donor_ids_basename = \"donor_ids.tsv\"\n",
    "sce_basename = \"data_sce.rds\"\n",
    "\n",
    "print('Reading donor_ids and sce files...')\n",
    "for (vl in vireo_libs_dated){\n",
    "    if (grepl('_', vl)){\n",
    "        vl = substr(vl, 8, nchar(vl))\n",
    "    }\n",
    "\n",
    "    donor_ids_path = paste(base_path, vl, donor_ids_basename, sep='/')\n",
    "    sce_path = paste(base_path, vl, sce_basename, sep='/')\n",
    "\n",
    "    tryCatch({\n",
    "        donor_ids = read.table(donor_ids_path, header = TRUE, sep = \"\\t\")\n",
    "        donor_ids_list[[vl]] = donor_ids\n",
    "        \n",
    "        sce = readRDS(sce_path)\n",
    "        sce_list[[vl]] = sce\n",
    "    }, error = function(e) {\n",
    "        message(\"Failed to read \", vl, \": \", e$message)\n",
    "    })\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "print(\"Reading Demuxlet donor assignments...\")\n",
    "demuxlet_donors_list = list()\n",
    "demuxlet_no_mt_donors_list = list()\n",
    "\n",
    "for(name in names(donor_ids_list)){\n",
    "    p=paste(base_path, name, 'demuxlet_demultiplexed.best', sep='/')\n",
    "    p_no_mt=paste(base_path, name, 'demuxlet_demultiplexed_no_mt.best', sep='/')\n",
    "    tryCatch({\n",
    "        demuxlet_donors_list[[name]] = read.table(p, header=TRUE)\n",
    "        demuxlet_no_mt_donors_list[[name]] = read.table(p_no_mt, header=TRUE)\n",
    "    }, error = function(e) {\n",
    "        message(\"Failed to read \", p, \": \", e$message)\n",
    "    })\n",
    "}\n",
    "\n",
    "\n",
    "batches = c(rep(1, 24), rep(2, 12))\n",
    "dapi_nurr = list(\n",
    "    'NURR', 'DAPI', 'NURR', 'DAPI',\n",
    "    'NURR', 'NURR', 'DAPI',\n",
    "    'NURR', 'NURR', 'DAPI',\n",
    "    'NURR', 'NURR', 'DAPI',\n",
    "    'NURR', 'NURR', 'DAPI',\n",
    "    'NURR', 'NURR', 'DAPI',\n",
    "    'NURR', 'NURR', 'DAPI',\n",
    "    'NURR', 'NURR', \n",
    "    'DAPI', 'DAPI', 'NURR',\n",
    "    'NURR', 'NURR', 'DAPI',\n",
    "    'DAPI', 'DAPI', 'DAPI',\n",
    "    'NURR', 'NURR', 'DAPI')\n",
    "names(batches) = names(donor_ids_list)\n",
    "names(dapi_nurr) = names(donor_ids_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719B</dt>\n",
       "\t\t<dd>'NURR'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719BD</dt>\n",
       "\t\t<dd>'DAPI'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719C</dt>\n",
       "\t\t<dd>'NURR'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719CD</dt>\n",
       "\t\t<dd>'DAPI'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719D1</dt>\n",
       "\t\t<dd>'NURR'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719D2</dt>\n",
       "\t\t<dd>'NURR'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719DD</dt>\n",
       "\t\t<dd>'DAPI'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719E1</dt>\n",
       "\t\t<dd>'NURR'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719E2</dt>\n",
       "\t\t<dd>'NURR'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719ED</dt>\n",
       "\t\t<dd>'DAPI'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719F1</dt>\n",
       "\t\t<dd>'NURR'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719F2</dt>\n",
       "\t\t<dd>'NURR'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719FD</dt>\n",
       "\t\t<dd>'DAPI'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719G1</dt>\n",
       "\t\t<dd>'NURR'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719G2</dt>\n",
       "\t\t<dd>'NURR'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719GD</dt>\n",
       "\t\t<dd>'DAPI'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719I1</dt>\n",
       "\t\t<dd>'NURR'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719I2</dt>\n",
       "\t\t<dd>'NURR'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719ID</dt>\n",
       "\t\t<dd>'DAPI'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719J1</dt>\n",
       "\t\t<dd>'NURR'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719J2</dt>\n",
       "\t\t<dd>'NURR'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719JD</dt>\n",
       "\t\t<dd>'DAPI'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719K1</dt>\n",
       "\t\t<dd>'NURR'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNiPoold230719K2</dt>\n",
       "\t\t<dd>'NURR'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNA8id230921A8</dt>\n",
       "\t\t<dd>'DAPI'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNA9id230921A9</dt>\n",
       "\t\t<dd>'DAPI'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNB8id230921B8</dt>\n",
       "\t\t<dd>'NURR'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNB9id230921B9</dt>\n",
       "\t\t<dd>'NURR'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNC8id230921C8</dt>\n",
       "\t\t<dd>'NURR'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNC9id230921C9</dt>\n",
       "\t\t<dd>'DAPI'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSND8id230921D8</dt>\n",
       "\t\t<dd>'DAPI'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSND9id230921D9</dt>\n",
       "\t\t<dd>'DAPI'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNE8id230921E8</dt>\n",
       "\t\t<dd>'DAPI'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNF8id230921F8</dt>\n",
       "\t\t<dd>'NURR'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNG8id230921G8</dt>\n",
       "\t\t<dd>'NURR'</dd>\n",
       "\t<dt>$pCalicoPDsHSrSNH8id230921H8</dt>\n",
       "\t\t<dd>'DAPI'</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719B] 'NURR'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719BD] 'DAPI'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719C] 'NURR'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719CD] 'DAPI'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719D1] 'NURR'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719D2] 'NURR'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719DD] 'DAPI'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719E1] 'NURR'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719E2] 'NURR'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719ED] 'DAPI'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719F1] 'NURR'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719F2] 'NURR'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719FD] 'DAPI'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719G1] 'NURR'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719G2] 'NURR'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719GD] 'DAPI'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719I1] 'NURR'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719I2] 'NURR'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719ID] 'DAPI'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719J1] 'NURR'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719J2] 'NURR'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719JD] 'DAPI'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719K1] 'NURR'\n",
       "\\item[\\$pCalicoPDsHSrSNiPoold230719K2] 'NURR'\n",
       "\\item[\\$pCalicoPDsHSrSNA8id230921A8] 'DAPI'\n",
       "\\item[\\$pCalicoPDsHSrSNA9id230921A9] 'DAPI'\n",
       "\\item[\\$pCalicoPDsHSrSNB8id230921B8] 'NURR'\n",
       "\\item[\\$pCalicoPDsHSrSNB9id230921B9] 'NURR'\n",
       "\\item[\\$pCalicoPDsHSrSNC8id230921C8] 'NURR'\n",
       "\\item[\\$pCalicoPDsHSrSNC9id230921C9] 'DAPI'\n",
       "\\item[\\$pCalicoPDsHSrSND8id230921D8] 'DAPI'\n",
       "\\item[\\$pCalicoPDsHSrSND9id230921D9] 'DAPI'\n",
       "\\item[\\$pCalicoPDsHSrSNE8id230921E8] 'DAPI'\n",
       "\\item[\\$pCalicoPDsHSrSNF8id230921F8] 'NURR'\n",
       "\\item[\\$pCalicoPDsHSrSNG8id230921G8] 'NURR'\n",
       "\\item[\\$pCalicoPDsHSrSNH8id230921H8] 'DAPI'\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$pCalicoPDsHSrSNiPoold230719B\n",
       ":   'NURR'\n",
       "$pCalicoPDsHSrSNiPoold230719BD\n",
       ":   'DAPI'\n",
       "$pCalicoPDsHSrSNiPoold230719C\n",
       ":   'NURR'\n",
       "$pCalicoPDsHSrSNiPoold230719CD\n",
       ":   'DAPI'\n",
       "$pCalicoPDsHSrSNiPoold230719D1\n",
       ":   'NURR'\n",
       "$pCalicoPDsHSrSNiPoold230719D2\n",
       ":   'NURR'\n",
       "$pCalicoPDsHSrSNiPoold230719DD\n",
       ":   'DAPI'\n",
       "$pCalicoPDsHSrSNiPoold230719E1\n",
       ":   'NURR'\n",
       "$pCalicoPDsHSrSNiPoold230719E2\n",
       ":   'NURR'\n",
       "$pCalicoPDsHSrSNiPoold230719ED\n",
       ":   'DAPI'\n",
       "$pCalicoPDsHSrSNiPoold230719F1\n",
       ":   'NURR'\n",
       "$pCalicoPDsHSrSNiPoold230719F2\n",
       ":   'NURR'\n",
       "$pCalicoPDsHSrSNiPoold230719FD\n",
       ":   'DAPI'\n",
       "$pCalicoPDsHSrSNiPoold230719G1\n",
       ":   'NURR'\n",
       "$pCalicoPDsHSrSNiPoold230719G2\n",
       ":   'NURR'\n",
       "$pCalicoPDsHSrSNiPoold230719GD\n",
       ":   'DAPI'\n",
       "$pCalicoPDsHSrSNiPoold230719I1\n",
       ":   'NURR'\n",
       "$pCalicoPDsHSrSNiPoold230719I2\n",
       ":   'NURR'\n",
       "$pCalicoPDsHSrSNiPoold230719ID\n",
       ":   'DAPI'\n",
       "$pCalicoPDsHSrSNiPoold230719J1\n",
       ":   'NURR'\n",
       "$pCalicoPDsHSrSNiPoold230719J2\n",
       ":   'NURR'\n",
       "$pCalicoPDsHSrSNiPoold230719JD\n",
       ":   'DAPI'\n",
       "$pCalicoPDsHSrSNiPoold230719K1\n",
       ":   'NURR'\n",
       "$pCalicoPDsHSrSNiPoold230719K2\n",
       ":   'NURR'\n",
       "$pCalicoPDsHSrSNA8id230921A8\n",
       ":   'DAPI'\n",
       "$pCalicoPDsHSrSNA9id230921A9\n",
       ":   'DAPI'\n",
       "$pCalicoPDsHSrSNB8id230921B8\n",
       ":   'NURR'\n",
       "$pCalicoPDsHSrSNB9id230921B9\n",
       ":   'NURR'\n",
       "$pCalicoPDsHSrSNC8id230921C8\n",
       ":   'NURR'\n",
       "$pCalicoPDsHSrSNC9id230921C9\n",
       ":   'DAPI'\n",
       "$pCalicoPDsHSrSND8id230921D8\n",
       ":   'DAPI'\n",
       "$pCalicoPDsHSrSND9id230921D9\n",
       ":   'DAPI'\n",
       "$pCalicoPDsHSrSNE8id230921E8\n",
       ":   'DAPI'\n",
       "$pCalicoPDsHSrSNF8id230921F8\n",
       ":   'NURR'\n",
       "$pCalicoPDsHSrSNG8id230921G8\n",
       ":   'NURR'\n",
       "$pCalicoPDsHSrSNH8id230921H8\n",
       ":   'DAPI'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$pCalicoPDsHSrSNiPoold230719B\n",
       "[1] \"NURR\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719BD\n",
       "[1] \"DAPI\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719C\n",
       "[1] \"NURR\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719CD\n",
       "[1] \"DAPI\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719D1\n",
       "[1] \"NURR\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719D2\n",
       "[1] \"NURR\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719DD\n",
       "[1] \"DAPI\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719E1\n",
       "[1] \"NURR\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719E2\n",
       "[1] \"NURR\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719ED\n",
       "[1] \"DAPI\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719F1\n",
       "[1] \"NURR\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719F2\n",
       "[1] \"NURR\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719FD\n",
       "[1] \"DAPI\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719G1\n",
       "[1] \"NURR\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719G2\n",
       "[1] \"NURR\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719GD\n",
       "[1] \"DAPI\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719I1\n",
       "[1] \"NURR\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719I2\n",
       "[1] \"NURR\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719ID\n",
       "[1] \"DAPI\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719J1\n",
       "[1] \"NURR\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719J2\n",
       "[1] \"NURR\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719JD\n",
       "[1] \"DAPI\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719K1\n",
       "[1] \"NURR\"\n",
       "\n",
       "$pCalicoPDsHSrSNiPoold230719K2\n",
       "[1] \"NURR\"\n",
       "\n",
       "$pCalicoPDsHSrSNA8id230921A8\n",
       "[1] \"DAPI\"\n",
       "\n",
       "$pCalicoPDsHSrSNA9id230921A9\n",
       "[1] \"DAPI\"\n",
       "\n",
       "$pCalicoPDsHSrSNB8id230921B8\n",
       "[1] \"NURR\"\n",
       "\n",
       "$pCalicoPDsHSrSNB9id230921B9\n",
       "[1] \"NURR\"\n",
       "\n",
       "$pCalicoPDsHSrSNC8id230921C8\n",
       "[1] \"NURR\"\n",
       "\n",
       "$pCalicoPDsHSrSNC9id230921C9\n",
       "[1] \"DAPI\"\n",
       "\n",
       "$pCalicoPDsHSrSND8id230921D8\n",
       "[1] \"DAPI\"\n",
       "\n",
       "$pCalicoPDsHSrSND9id230921D9\n",
       "[1] \"DAPI\"\n",
       "\n",
       "$pCalicoPDsHSrSNE8id230921E8\n",
       "[1] \"DAPI\"\n",
       "\n",
       "$pCalicoPDsHSrSNF8id230921F8\n",
       "[1] \"NURR\"\n",
       "\n",
       "$pCalicoPDsHSrSNG8id230921G8\n",
       "[1] \"NURR\"\n",
       "\n",
       "$pCalicoPDsHSrSNH8id230921H8\n",
       "[1] \"DAPI\"\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dapi_nurr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "if (READ_RAW_READS) {\n",
    "\n",
    "    print(\"Reading raw data...\")\n",
    "    raw_data_list = list()\n",
    "    raw_col_data_list = list()\n",
    "    for (name in names(donor_ids_list)){\n",
    "        raw_base_path = \"/mnt/accessory/seq_data/calico_raw\" \n",
    "        raw_file_name = \"raw_feature_bc_matrix.rds\"\n",
    "        raw_file_path = paste(raw_base_path, name, raw_file_name, sep='/')\n",
    "\n",
    "        #raw_data_dcg = sum_duplicate_rownames_of_dcg_matrix(readRDS(raw_file_path))\n",
    "        raw_data_dcg = readRDS(raw_file_path)\n",
    "        raw_data_list[[name]] = raw_data_dcg\n",
    "        cd = data.frame(\n",
    "            row.names=colnames(raw_data_dcg),\n",
    "            nUMI=colSums(raw_data_dcg),\n",
    "            log10_nUMI = log10(colSums(raw_data_dcg) + 1),\n",
    "            nGene=colSums(raw_data_dcg > 0),\n",
    "            is_in_filtered=colnames(raw_data_dcg) %in% colnames(sce_list[[name]])\n",
    "        )\n",
    "        raw_col_data_list[[name]] = cd\n",
    "        saveRDS(cd, paste(raw_base_path, name, 'raw_feature_bc_matrix_col_data.rds', sep='/'))\n",
    "    }\n",
    "}\n",
    "\n",
    "cd_list = list()\n",
    "for (name in names(donor_ids_list)){\n",
    "    cd = as.data.frame(colData(sce_list[[name]]))\n",
    "    cd$log10_nUMI = log10(cd$nUMI + 1)\n",
    "    cd$log10_nGene = log10(cd$nGene + 1)\n",
    "    cd$log10_nRead = log10(cd$nRead + 1)\n",
    "    cd_list[[name]] = cd\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# TAKES 20 MINUTES\n",
    "# molecule_info_list = list()\n",
    "\n",
    "# for (name in names(donor_ids_list)){\n",
    "#     molecule_info_list[[name]] = readRDS(paste(base_path, name, 'outs/molecule_info.rds', sep='/'))\n",
    "# }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "getAndPlotConfusionMatrix = function(\n",
    "    merged_df,\n",
    "    name,\n",
    "    col_x,\n",
    "    col_y,\n",
    "    non_assignable_categories = c('doublet', 'unassigned'),\n",
    "    xlab=NULL,\n",
    "    ylab=NULL,\n",
    "    title=NULL,\n",
    "    fig_filename=NULL\n",
    "){\n",
    "    # TODO: add norm args for either cols, rows, or the whole matrix\n",
    "\n",
    "    # Create the confusion matrix\n",
    "    x_factors = factor(merged_df[[col_x]])\n",
    "    y_factors = factor(merged_df[[col_y]])\n",
    "\n",
    "    # \"data\" and \"reference\" are misnomers; we're just comparing the two methods\n",
    "    # \"reference\" labels displayed on X-Axis, \"data\" on Y-Axis\n",
    "    # Below, the xlab MUST correspond to the \"reference\", and the ylab to the \"data\"\n",
    "    conf_mat <- confusionMatrix(data=y_factors, reference=x_factors)\n",
    "\n",
    "    # Get the table of raw counts and convert it to proportions\n",
    "    conf_mat_table <- conf_mat$table\n",
    "    prop_table_col_norm <- conf_mat_table #apply(conf_mat_table, 2, function(x) x / sum(x)) # divide each column by its sum. Need to be specific about axis bc matrix is square\n",
    "\n",
    "    if (is.null(xlab)) { xlab = col_x }\n",
    "    if (is.null(ylab)) { ylab = col_y }\n",
    "    if (is.null(title)){\n",
    "        title = paste(\"Confusion Matrix for library\", name, '\\n', ylab, \"vs:\", xlab, sep=' ')\n",
    "    }\n",
    "\n",
    "    merged_df_doubly_assignable = merged_df[\n",
    "        !(merged_df[[col_x]] %in% non_assignable_categories) & \n",
    "        !(merged_df[[col_y]] %in% non_assignable_categories),]\n",
    "\n",
    "    conf_mat_doubly_assignable = confusionMatrix(\n",
    "        data=factor(merged_df_doubly_assignable[[col_y]]), \n",
    "        reference=factor(merged_df_doubly_assignable[[col_x]])\n",
    "    )\n",
    "\n",
    "    title = paste0(title,                 \n",
    "        '\\nFrac of Cells with Same Predictions: ', round(conf_mat$overall['Accuracy'], 3),\n",
    "        '\\nFrac Assignable by both methods: ', round(dim(merged_df_doubly_assignable)[1] / dim(merged_df)[1], 2),\n",
    "        '\\nFrac of Doubly Assignable Cells with Same Predictions: ', round(conf_mat_doubly_assignable$overall['Accuracy'], 3)\n",
    "    )\n",
    "\n",
    "    # Call the function to plot the heatmap\n",
    "    # the xlab MUST correspond to the reference, and the ylab to the data\n",
    "    plotConfusionHeatmap(\n",
    "        prop_table_col_norm, \n",
    "        xlab=xlab, \n",
    "        ylab=ylab, \n",
    "        title=title,\n",
    "        fig_filename=fig_filename\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "source(\"~/sc-online/plot.R\")\n",
    "\n",
    "# For every library, plot the following\n",
    "\n",
    "# 1. Display a df of\n",
    "#   (a) num cells\n",
    "#   (b) Med num genes\n",
    "#   (c) Med num UMIs\n",
    "#   (d) Min num UMIs\n",
    "#   (e) Med Reads / UMI\n",
    "\n",
    "# 2. Knee Plot\n",
    "# 3. Hist of Doublet Probabilities\n",
    "# 4. Hist of nUMI colored by Assignable\n",
    "# 5. Hist of Reads / UMI colored by assignable\n",
    "# 6. Scatter of Pct Intronic vs nUMI, colored by assignable\n",
    "# 7. Vireo / Demuxlet Confusion Matrixes\n",
    "# 8. Vireo / Demuxlet Confusion Matrixes -- Doubly Assignable Cells Only\n",
    "\n",
    "plot_base_dir = '/mnt/accessory/analysis/plots'\n",
    "\n",
    "for (name in names(donor_ids_list[1:1])){\n",
    "\n",
    "    print(name)\n",
    "\n",
    "    vireo_donors = donor_ids_list[[name]]\n",
    "    sce = sce_list[[name]]\n",
    "    demuxlet_donors = demuxlet_donors_list[[name]]\n",
    "    raw_data = raw_data_list[[name]]\n",
    "    raw_col_data = raw_col_data_list[[name]]\n",
    "    #molecule_info = molecule_info_list[[name]]\n",
    "\n",
    "    df_plot_path = file.path(plot_base_dir, paste0(name, '_1_df_plot.png'))\n",
    "    knee_plot_path = file.path(plot_base_dir, paste0(name, '_2_knee_plot.png'))\n",
    "    doublet_prob_hist_path = file.path(plot_base_dir, paste0(name, '_3_doublet_prob_hist.png'))\n",
    "    num_umi_hist_path = file.path(plot_base_dir, paste0(name, '_4_num_umi_hist.png'))\n",
    "    reads_per_umi_hist_path = file.path(plot_base_dir, paste0(name, '_5_reads_per_umi_hist.png'))\n",
    "    pct_intronic_vs_num_umi_path = file.path(plot_base_dir, paste0(name, '_6_pct_intronic_vs_num_umi.png'))\n",
    "    vireo_confusion_matrix_path = file.path(plot_base_dir, paste0(name, '_7_vireo_demuxlet_confusion_matrix.png'))\n",
    "\n",
    "    raw_col_data$log10_nUMI = log10(raw_col_data$nUMI + 1)\n",
    "\n",
    "    cd = as.data.frame(colData(sce))\n",
    "    cd$log10_nUMI = log10(cd$nUMI + 1)\n",
    "    cd$log10_nGene = log10(cd$nGene + 1)\n",
    "    cd$log10_nRead = log10(cd$nRead + 1)\n",
    "    cd$prob_max = cd$prob_donor # this is just for compatibility with other functions\n",
    "    cd$reads_per_umi = cd$nRead / cd$nUMI\n",
    "\n",
    "    cd_singlet = cd[cd$prob_doublet < (1/3), ]\n",
    "\n",
    "    # build a df that is the merger of demuxlet and vireo, and change column names for ease of use\n",
    "    merged_df = merge(demuxlet_donors, vireo_donors, by.x='BARCODE', by.y='cell', all.x=TRUE, all.y=TRUE)\n",
    "    rownames(merged_df) = merged_df$BARCODE\n",
    "    \n",
    "    merged_df['is_singlet_vireo'] = as.numeric(merged_df$prob_doublet < (1/3))\n",
    "    merged_df['is_singlet_demuxlet'] = as.numeric(substr(merged_df$BEST, 1, 3) == 'SNG')\n",
    "\n",
    "    merged_df['prob_donor_vireo'] = merged_df$prob_max\n",
    "    merged_df['prob_donor_demuxlet'] = merged_df$PRB.SNG1\n",
    "\n",
    "    # NOTE: THERE ARE CASES WHEN VIREO'S `DONOR_ID` IS 'UNASSIGNED' EVEN WHEN `PROB_MAX` >= 0.9\n",
    "    # IN CASES WITH FEWER THAN 10 SNPS. BEST_SINGLET IGNORES THIS CUTOFF\n",
    "    merged_df['id_donor_vireo'] = merged_df$donor_id\n",
    "    merged_df$id_donor_demuxlet = ifelse(substr(merged_df$BEST, 1, 3) == 'SNG', \n",
    "                                      ifelse(merged_df$PRB.SNG1 >= 0.9, \n",
    "                                             substr(merged_df$BEST, 5, nchar(merged_df$BEST)), \n",
    "                                             'unassigned'),\n",
    "                                      ifelse(substr(merged_df$BEST, 1, 3) == 'DBL', \n",
    "                                             'doublet', \n",
    "                                             'unassigned'))\n",
    "\n",
    "    merged_df['library'] = name\n",
    "\n",
    "    # summary df to plot\n",
    "    summary_df = data.frame(\n",
    "        'Library' = name,\n",
    "        'Sort' = dapi_nurr[[name]],\n",
    "        'Batch' = batches[[name]],\n",
    "        'Num_Cells' = nrow(cd),\n",
    "        'Frac_Assignable_Vireo' = round(getFracAssignableVireo(vireo_donors), 2),\n",
    "        'Frac_Assignable_Demuxlet' = round(getFracAssignableDemuxlet(demuxlet_donors), 2),\n",
    "        \n",
    "        'Min_nReads' = min(cd$nRead),\n",
    "        'Median_nReads' = median(cd$nRead),\n",
    "        'Mean_nReads' = round(mean(cd$nRead), 2),\n",
    "\n",
    "        'Min_nUMI' = min(cd$nUMI),\n",
    "        'Median_nUMI' = median(cd$nUMI),\n",
    "        'Mean_nUMI' = round(mean(cd$nUMI), 2),\n",
    "\n",
    "        'Min_nGenes' = min(cd$nGene),\n",
    "        'Median_nGenes' = median(cd$nGene),\n",
    "        'Mean_nGenes' = round(mean(cd$nGene), 2),\n",
    "\n",
    "        'Min_Reads_per_UMI' = round(min(cd$reads_per_umi), 2),\n",
    "        'Median_Reads_per_UMI' = round(median(cd$reads_per_umi), 2),\n",
    "        'Mean_Reads_per_UMI' = round(mean(cd$reads_per_umi), 2),\n",
    "\n",
    "        'Min_frac_intronic' = round(min(cd$pct_intronic), 2),\n",
    "        'Median_frac_intronic' = round(median(cd$pct_intronic), 2),\n",
    "        'Mean_frac_intronic' = round(mean(cd$pct_intronic), 2),\n",
    "        'Max_frac_intronic' = round(max(cd$pct_intronic), 2),\n",
    "\n",
    "        'Frac_Singlet_Vireo' = round(nrow(cd_singlet) / nrow(cd), 2),\n",
    "        'Frac_Singlet_Demuxlet' =  round(sum(substr(demuxlet_donors$BEST, 1, 3) == 'SNG') / nrow(demuxlet_donors), 2),\n",
    "        'Frac_Singlet_Vireo_Assignable' = round(getFracAssignableVireo(vireo_donors[vireo_donors$prob_doublet < (1/3), ]), 2),\n",
    "        'Frac_Singlet_Demuxlet_Assignable' = round(getFracAssignableDemuxlet(demuxlet_donors[substr(demuxlet_donors$BEST, 1, 3) == 'SNG',]), 2)\n",
    "    )\n",
    "\n",
    "    \n",
    "\n",
    "    # 1. display DF\n",
    "    plotDfSummary(\n",
    "        df=as.data.frame(t(summary_df)),\n",
    "        fig_filename=df_plot_path\n",
    "    )\n",
    "\n",
    "    # 2. Knee\n",
    "    plotKneeSingle(\n",
    "        df=raw_col_data,\n",
    "        name=name,\n",
    "        title=paste0(\n",
    "            'Knee Plot: ', name,\n",
    "            '\\nNum Cells: ', sum(raw_col_data$is_in_filtered),\n",
    "            '\\nFraction of Cells Assignable: ', round(getFracAssignableVireo(cd), 2)),\n",
    "        ylim=c(0.5, 6),\n",
    "        clim=c(.15, .8),\n",
    "        \n",
    "        fig_filename=knee_plot_path\n",
    "    )\n",
    "\n",
    "    # 3. Doublet Prob Hist\n",
    "    plotHistColorSingle(\n",
    "        df=vireo_donors,\n",
    "        name=name,\n",
    "        plot_col='prob_doublet',\n",
    "        color_col='prob_max',\n",
    "        title=paste0(\n",
    "            'Vireo Doublet Probability: ', name,\n",
    "            \"\\nFrac Predicted Doublet: \", round(sum(vireo_donors$prob_doublet >= (1/3)) / nrow(vireo_donors), 2)), \n",
    "        xlim=c(0, 1),\n",
    "        ylim=c(0, 1),\n",
    "        bin_width=0.1,\n",
    "        fig_filename=doublet_prob_hist_path\n",
    "    )\n",
    "\n",
    "    # 4. Num UMI Hist\n",
    "    plotHistColorSingle(\n",
    "        df=cd_singlet,\n",
    "        name=name,\n",
    "        plot_col='log10_nUMI',\n",
    "        color_col='prob_donor',\n",
    "        bin_width=0.2,\n",
    "    xlim=c(2.6, 5),\n",
    "    ylim=c(0, 0.5),\n",
    "        title=paste0(\n",
    "            'Singlet Log10(nUMI) Histogram -- ', name,\n",
    "            '\\nFraction of Singlets Assignable: ', round(getFracAssignableVireo(cd_singlet), 2),\n",
    "            '\\nUMI -- Min: ', min(cd_singlet$nUMI), ' -- Median: ', median(cd_singlet$nUMI), ' -- Mean: ', mean(cd_singlet$nUMI)),\n",
    "        fig_filename=num_umi_hist_path\n",
    "    )\n",
    "\n",
    "    # 5. Reads / UMI Hist\n",
    "    plotHistColorSingle(\n",
    "        df=cd_singlet,\n",
    "        name=name,\n",
    "        plot_col='reads_per_umi',\n",
    "        color_col='prob_donor',\n",
    "        xlim=c(round(min(cd_singlet$reads_per_umi)), round(max(cd_singlet$reads_per_umi))),\n",
    "        bin_width=round(max(cd_singlet$reads_per_umi) - min(cd_singlet$reads_per_umi)) / 20,\n",
    "        ylim=c(0, 0.5),\n",
    "        title=paste0(\n",
    "            'Singlet nReads/nUMI Histogram -- ', name,\n",
    "            '\\nFraction of Singlets Assignable: ', round(getFracAssignableVireo(cd_singlet), 2),\n",
    "            '\\nUMI -- Min: ', round(min(cd_singlet$reads_per_umi), 1), \n",
    "            ' -- Median: ', round(median(cd_singlet$reads_per_umi), 1),\n",
    "            ' -- Mean: ', round(mean(cd_singlet$reads_per_umi), 1)),\n",
    "        fig_filename=reads_per_umi_hist_path\n",
    "    )\n",
    "\n",
    "    #6. Pct Intronic vs Num UMI Scatter\n",
    "    plotScatterColorSingle(\n",
    "        df=cd_singlet,\n",
    "        name=name,\n",
    "        x_col='log10_nUMI',\n",
    "        y_col='pct_intronic',\n",
    "        color_col='prob_donor',\n",
    "        xlim=c(2.5, 6),\n",
    "        ylim=c(0, 1),\n",
    "        plot_height=8,\n",
    "        plot_width=8,\n",
    "        fig_filename=pct_intronic_vs_num_umi_path\n",
    "    )\n",
    "\n",
    "    # # 7. Confusion Matrix\n",
    "    getAndPlotConfusionMatrix(\n",
    "        merged_df=merged_df,\n",
    "        name=name,\n",
    "        col_x='id_donor_vireo', col_y='id_donor_demuxlet',\n",
    "        xlab='Vireo', ylab='Demuxlet', \n",
    "        title=paste0(\n",
    "            name, ' -- Batch ', batches[[name]], ' -- Sort: ', dapi_nurr[[name]],\n",
    "            '\\nConfusion Matrix:', \n",
    "                '\\nFrac Assignable Vireo: ', round(getFracAssignableVireo(merged_df), 2),\n",
    "                ' -- Frac Assignable Demuxlet: ', round(getFracAssignableDemuxlet(merged_df), 2)\n",
    "            ),\n",
    "        fig_filename=vireo_confusion_matrix_path\n",
    "    )\n",
    "    \n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
