{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: EnsDb.Hsapiens.v86\n",
      "\n",
      "Loading required package: ensembldb\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: GenomicFeatures\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Possible Ensembl SSL connectivity problems detected.\n",
      "Please see the 'Connection Troubleshooting' section of the biomaRt vignette\n",
      "vignette('accessing_ensembl', package = 'biomaRt')Error in curl::curl_fetch_memory(url, handle = handle) : \n",
      "  SSL peer certificate or SSH remote key was not OK: [uswest.ensembl.org] SSL certificate problem: certificate has expired\n",
      "\n",
      "\n",
      "Loading required package: AnnotationFilter\n",
      "\n",
      "\n",
      "Attaching package: 'ensembldb'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "Loading required package: EnsDb.Mmusculus.v79\n",
      "\n",
      "Loading required package: SingleCellExperiment\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: 'matrixStats'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:Biobase':\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'MatrixGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:matrixStats':\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'dplyr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:matrixStats':\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:ensembldb':\n",
      "\n",
      "    filter, select\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationDbi':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:GenomicRanges':\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomeInfoDb':\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:IRanges':\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:S4Vectors':\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:BiocGenerics':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,\n",
      "which was just loaded, were retired in October 2023.\n",
      "Please refer to R-spatial evolution reports for details, especially\n",
      "https://r-spatial.org/r/2023/05/15/evolution4.html.\n",
      "It may be desirable to make the sf package available;\n",
      "package maintainers should consider adding sf to Suggests:.\n",
      "\n",
      "Attaching SeuratObject\n",
      "\n",
      "\n",
      "Attaching package: 'Seurat'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:SummarizedExperiment':\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'tidyr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:S4Vectors':\n",
      "\n",
      "    expand\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "source(\"~/sc-online/extraGenes.R\")\n",
    "source(\"~/sc-online/myExpSetCreatorFn.R\")\n",
    "source(\"~/sc-online/utils.R\")\n",
    "\n",
    "library(\"DropletUtils\")\n",
    "library(dplyr)\n",
    "library(lisi)\n",
    "library(rhdf5)\n",
    "library(Seurat)\n",
    "library(tidyr)\n",
    "\n",
    "\n",
    "# This magic number is extremely important; \n",
    "# it's the probability above which we will take the max of Vireo's predictions of the likely donor \n",
    "DONOR_ASSIGNMENT_CUTOFF = 0.3\n",
    "\n",
    "SAVE_H5_TO_RDS=FALSE\n",
    "\n",
    "base_path = \"~/seq_data/pd\"\n",
    "mol_path  = \"outs/molecule_info.h5\"\n",
    "umi_path = \"outs/filtered_feature_bc_matrix.h5\"\n",
    "donor_ids_path = \"donor_ids.tsv\"\n",
    "prob_singlet_path = \"prob_singlet.tsv\"\n",
    "prob_doublet_path = \"prob_doublet.tsv\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'bam_path'</li><li>'BCLPath'</li><li>'BICCN.ALREADY.UPLOADED'</li><li>'BICCN.TO.UPLOAD'</li><li>'chemistry_10x'</li><li>'date.YYMMDD.'</li><li>'demultiplex'</li><li>'donor_id'</li><li>'donor_list'</li><li>'email'</li><li>'Estimated.Number..of.Cells'</li><li>'FACS'</li><li>'fastq_path'</li><li>'flowcell'</li><li>'Fraction.Reads.in.Cells'</li><li>'genotype_VCF_path'</li><li>'IlluminaPlatform'</li><li>'Index'</li><li>'Lane'</li><li>'library'</li><li>'library_path'</li><li>'Matches.library.name.formula'</li><li>'Mean.Reads..per.Cell'</li><li>'Median.Genes..per.Cell'</li><li>'Median.UMI.Counts.per.Cell'</li><li>'Nuclei.Extraction.Protocol'</li><li>'num_expected_cells'</li><li>'Number.of..Reads'</li><li>'Pass_qc'</li><li>'PlateID'</li><li>'project_name'</li><li>'Q30.Bases.in.Barcode'</li><li>'Q30.Bases.in.RNA.Read'</li><li>'Q30.Bases.in.Sample.Index'</li><li>'Q30.Bases.in.UMI'</li><li>'Read.first..Summary.HTML.Link..copy.and.paste.into.new.tab.in.Chrome....must.have.mounted.Broad.cifs.drive.locally..Mac.link.'</li><li>'Reads.Mapped.Antisense.to.Gene'</li><li>'Reads.Mapped.Confidently.to.Exonic.Regions'</li><li>'Reads.Mapped.Confidently.to.Genome'</li><li>'Reads.Mapped.Confidently.to.Intergenic.Regions'</li><li>'Reads.Mapped.Confidently.to.Intronic.Regions'</li><li>'Reads.Mapped.Confidently.to.Transcriptome'</li><li>'Reads.Mapped.to.Genome'</li><li>'region'</li><li>'RNA.Prep.Method'</li><li>'Sequencing..Saturation'</li><li>'Sex'</li><li>'species'</li><li>'sub_region.tube_id'</li><li>'suffix..optional.'</li><li>'Total.Genes.Detected'</li><li>'transcriptome'</li><li>'Uploaded_NEMO'</li><li>'Valid.Barcodes'</li><li>'X'</li><li>'X.1'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'bam\\_path'\n",
       "\\item 'BCLPath'\n",
       "\\item 'BICCN.ALREADY.UPLOADED'\n",
       "\\item 'BICCN.TO.UPLOAD'\n",
       "\\item 'chemistry\\_10x'\n",
       "\\item 'date.YYMMDD.'\n",
       "\\item 'demultiplex'\n",
       "\\item 'donor\\_id'\n",
       "\\item 'donor\\_list'\n",
       "\\item 'email'\n",
       "\\item 'Estimated.Number..of.Cells'\n",
       "\\item 'FACS'\n",
       "\\item 'fastq\\_path'\n",
       "\\item 'flowcell'\n",
       "\\item 'Fraction.Reads.in.Cells'\n",
       "\\item 'genotype\\_VCF\\_path'\n",
       "\\item 'IlluminaPlatform'\n",
       "\\item 'Index'\n",
       "\\item 'Lane'\n",
       "\\item 'library'\n",
       "\\item 'library\\_path'\n",
       "\\item 'Matches.library.name.formula'\n",
       "\\item 'Mean.Reads..per.Cell'\n",
       "\\item 'Median.Genes..per.Cell'\n",
       "\\item 'Median.UMI.Counts.per.Cell'\n",
       "\\item 'Nuclei.Extraction.Protocol'\n",
       "\\item 'num\\_expected\\_cells'\n",
       "\\item 'Number.of..Reads'\n",
       "\\item 'Pass\\_qc'\n",
       "\\item 'PlateID'\n",
       "\\item 'project\\_name'\n",
       "\\item 'Q30.Bases.in.Barcode'\n",
       "\\item 'Q30.Bases.in.RNA.Read'\n",
       "\\item 'Q30.Bases.in.Sample.Index'\n",
       "\\item 'Q30.Bases.in.UMI'\n",
       "\\item 'Read.first..Summary.HTML.Link..copy.and.paste.into.new.tab.in.Chrome....must.have.mounted.Broad.cifs.drive.locally..Mac.link.'\n",
       "\\item 'Reads.Mapped.Antisense.to.Gene'\n",
       "\\item 'Reads.Mapped.Confidently.to.Exonic.Regions'\n",
       "\\item 'Reads.Mapped.Confidently.to.Genome'\n",
       "\\item 'Reads.Mapped.Confidently.to.Intergenic.Regions'\n",
       "\\item 'Reads.Mapped.Confidently.to.Intronic.Regions'\n",
       "\\item 'Reads.Mapped.Confidently.to.Transcriptome'\n",
       "\\item 'Reads.Mapped.to.Genome'\n",
       "\\item 'region'\n",
       "\\item 'RNA.Prep.Method'\n",
       "\\item 'Sequencing..Saturation'\n",
       "\\item 'Sex'\n",
       "\\item 'species'\n",
       "\\item 'sub\\_region.tube\\_id'\n",
       "\\item 'suffix..optional.'\n",
       "\\item 'Total.Genes.Detected'\n",
       "\\item 'transcriptome'\n",
       "\\item 'Uploaded\\_NEMO'\n",
       "\\item 'Valid.Barcodes'\n",
       "\\item 'X'\n",
       "\\item 'X.1'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'bam_path'\n",
       "2. 'BCLPath'\n",
       "3. 'BICCN.ALREADY.UPLOADED'\n",
       "4. 'BICCN.TO.UPLOAD'\n",
       "5. 'chemistry_10x'\n",
       "6. 'date.YYMMDD.'\n",
       "7. 'demultiplex'\n",
       "8. 'donor_id'\n",
       "9. 'donor_list'\n",
       "10. 'email'\n",
       "11. 'Estimated.Number..of.Cells'\n",
       "12. 'FACS'\n",
       "13. 'fastq_path'\n",
       "14. 'flowcell'\n",
       "15. 'Fraction.Reads.in.Cells'\n",
       "16. 'genotype_VCF_path'\n",
       "17. 'IlluminaPlatform'\n",
       "18. 'Index'\n",
       "19. 'Lane'\n",
       "20. 'library'\n",
       "21. 'library_path'\n",
       "22. 'Matches.library.name.formula'\n",
       "23. 'Mean.Reads..per.Cell'\n",
       "24. 'Median.Genes..per.Cell'\n",
       "25. 'Median.UMI.Counts.per.Cell'\n",
       "26. 'Nuclei.Extraction.Protocol'\n",
       "27. 'num_expected_cells'\n",
       "28. 'Number.of..Reads'\n",
       "29. 'Pass_qc'\n",
       "30. 'PlateID'\n",
       "31. 'project_name'\n",
       "32. 'Q30.Bases.in.Barcode'\n",
       "33. 'Q30.Bases.in.RNA.Read'\n",
       "34. 'Q30.Bases.in.Sample.Index'\n",
       "35. 'Q30.Bases.in.UMI'\n",
       "36. 'Read.first..Summary.HTML.Link..copy.and.paste.into.new.tab.in.Chrome....must.have.mounted.Broad.cifs.drive.locally..Mac.link.'\n",
       "37. 'Reads.Mapped.Antisense.to.Gene'\n",
       "38. 'Reads.Mapped.Confidently.to.Exonic.Regions'\n",
       "39. 'Reads.Mapped.Confidently.to.Genome'\n",
       "40. 'Reads.Mapped.Confidently.to.Intergenic.Regions'\n",
       "41. 'Reads.Mapped.Confidently.to.Intronic.Regions'\n",
       "42. 'Reads.Mapped.Confidently.to.Transcriptome'\n",
       "43. 'Reads.Mapped.to.Genome'\n",
       "44. 'region'\n",
       "45. 'RNA.Prep.Method'\n",
       "46. 'Sequencing..Saturation'\n",
       "47. 'Sex'\n",
       "48. 'species'\n",
       "49. 'sub_region.tube_id'\n",
       "50. 'suffix..optional.'\n",
       "51. 'Total.Genes.Detected'\n",
       "52. 'transcriptome'\n",
       "53. 'Uploaded_NEMO'\n",
       "54. 'Valid.Barcodes'\n",
       "55. 'X'\n",
       "56. 'X.1'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"bam_path\"                                                                                                                     \n",
       " [2] \"BCLPath\"                                                                                                                      \n",
       " [3] \"BICCN.ALREADY.UPLOADED\"                                                                                                       \n",
       " [4] \"BICCN.TO.UPLOAD\"                                                                                                              \n",
       " [5] \"chemistry_10x\"                                                                                                                \n",
       " [6] \"date.YYMMDD.\"                                                                                                                 \n",
       " [7] \"demultiplex\"                                                                                                                  \n",
       " [8] \"donor_id\"                                                                                                                     \n",
       " [9] \"donor_list\"                                                                                                                   \n",
       "[10] \"email\"                                                                                                                        \n",
       "[11] \"Estimated.Number..of.Cells\"                                                                                                   \n",
       "[12] \"FACS\"                                                                                                                         \n",
       "[13] \"fastq_path\"                                                                                                                   \n",
       "[14] \"flowcell\"                                                                                                                     \n",
       "[15] \"Fraction.Reads.in.Cells\"                                                                                                      \n",
       "[16] \"genotype_VCF_path\"                                                                                                            \n",
       "[17] \"IlluminaPlatform\"                                                                                                             \n",
       "[18] \"Index\"                                                                                                                        \n",
       "[19] \"Lane\"                                                                                                                         \n",
       "[20] \"library\"                                                                                                                      \n",
       "[21] \"library_path\"                                                                                                                 \n",
       "[22] \"Matches.library.name.formula\"                                                                                                 \n",
       "[23] \"Mean.Reads..per.Cell\"                                                                                                         \n",
       "[24] \"Median.Genes..per.Cell\"                                                                                                       \n",
       "[25] \"Median.UMI.Counts.per.Cell\"                                                                                                   \n",
       "[26] \"Nuclei.Extraction.Protocol\"                                                                                                   \n",
       "[27] \"num_expected_cells\"                                                                                                           \n",
       "[28] \"Number.of..Reads\"                                                                                                             \n",
       "[29] \"Pass_qc\"                                                                                                                      \n",
       "[30] \"PlateID\"                                                                                                                      \n",
       "[31] \"project_name\"                                                                                                                 \n",
       "[32] \"Q30.Bases.in.Barcode\"                                                                                                         \n",
       "[33] \"Q30.Bases.in.RNA.Read\"                                                                                                        \n",
       "[34] \"Q30.Bases.in.Sample.Index\"                                                                                                    \n",
       "[35] \"Q30.Bases.in.UMI\"                                                                                                             \n",
       "[36] \"Read.first..Summary.HTML.Link..copy.and.paste.into.new.tab.in.Chrome....must.have.mounted.Broad.cifs.drive.locally..Mac.link.\"\n",
       "[37] \"Reads.Mapped.Antisense.to.Gene\"                                                                                               \n",
       "[38] \"Reads.Mapped.Confidently.to.Exonic.Regions\"                                                                                   \n",
       "[39] \"Reads.Mapped.Confidently.to.Genome\"                                                                                           \n",
       "[40] \"Reads.Mapped.Confidently.to.Intergenic.Regions\"                                                                               \n",
       "[41] \"Reads.Mapped.Confidently.to.Intronic.Regions\"                                                                                 \n",
       "[42] \"Reads.Mapped.Confidently.to.Transcriptome\"                                                                                    \n",
       "[43] \"Reads.Mapped.to.Genome\"                                                                                                       \n",
       "[44] \"region\"                                                                                                                       \n",
       "[45] \"RNA.Prep.Method\"                                                                                                              \n",
       "[46] \"Sequencing..Saturation\"                                                                                                       \n",
       "[47] \"Sex\"                                                                                                                          \n",
       "[48] \"species\"                                                                                                                      \n",
       "[49] \"sub_region.tube_id\"                                                                                                           \n",
       "[50] \"suffix..optional.\"                                                                                                            \n",
       "[51] \"Total.Genes.Detected\"                                                                                                         \n",
       "[52] \"transcriptome\"                                                                                                                \n",
       "[53] \"Uploaded_NEMO\"                                                                                                                \n",
       "[54] \"Valid.Barcodes\"                                                                                                               \n",
       "[55] \"X\"                                                                                                                            \n",
       "[56] \"X.1\"                                                                                                                          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "expt_log = read.csv('~/expt_log.csv')\n",
    "sort(names(expt_log))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>49</li><li>15</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 49\n",
       "\\item 15\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 49\n",
       "2. 15\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 49 15"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# only keep project names that are either PD or CalicoPD\n",
    "expt_log_pd = expt_log[expt_log$project_name %in% c('PD', 'CalicoPD'),]\n",
    "\n",
    "# Mehrdad started collecting data on 221126\n",
    "expt_log_pd = expt_log_pd[expt_log_pd$date.YYMMDD. > 221125,]\n",
    "\n",
    "# Homo Sapiens\n",
    "expt_log_pd_human = expt_log_pd[expt_log_pd$species == 'HS',]\n",
    "\n",
    "# Substantia Nigra\n",
    "expt_log_pd_human_sn = expt_log_pd_human[expt_log_pd_human$region == 'SN',]\n",
    "\n",
    "#subset columns\n",
    "cols2keep = c(\n",
    "    'library',\n",
    "    'region',\n",
    "    'species',\n",
    "    'sub_region.tube_id',\n",
    "    'donor_id',\n",
    "    'date.YYMMDD.',\n",
    "    'flowcell',\n",
    "    'BCLPath',\n",
    "    'IlluminaPlatform',\n",
    "    'num_expected_cells',\n",
    "    'transcriptome',\n",
    "    'demultiplex',\n",
    "    'genotype_VCF_path',\n",
    "    'donor_list',\n",
    "    'Lane'\n",
    ")\n",
    "expt_log_pd_human_sn = expt_log_pd_human_sn[,cols2keep]\n",
    "\n",
    "# keep experiments that HAVE been demultiplexed\n",
    "expt_log_pd_human_sn_demult = expt_log_pd_human_sn[expt_log_pd_human_sn$demultiplex == TRUE,]\n",
    "dim(expt_log_pd_human_sn_demult)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>17</li><li>15</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 17\n",
       "\\item 15\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 17\n",
       "2. 15\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 17 15"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# of these experiments, want to analyze those with both molecular counts and with demultiplexing data\n",
    "find_outs_command = \"find ~/seq_data/pd -mindepth 2 -maxdepth 2 -type d -name outs -exec dirname {} \\\\;\"\n",
    "counts_list = system(find_outs_command, intern=T)\n",
    "\n",
    "find_demult_command = \"find ~/seq_data/pd -mindepth 2 -maxdepth 2 -type f -name prob_doublet.tsv -exec dirname {} \\\\;\"\n",
    "demult_list = system(find_demult_command, intern=T)\n",
    "\n",
    "# get the intersection of the two lists\n",
    "counts_and_demult_list = basename(counts_list[basename(counts_list) %in% basename(demult_list)])\n",
    "\n",
    "# only keep experiments with both \n",
    "expt_log_pd_human_sn_demult = expt_log_pd_human_sn_demult[expt_log_pd_human_sn_demult$library %in% counts_and_demult_list,]\n",
    "dim(expt_log_pd_human_sn_demult)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "if(SAVE_H5_TO_RDS){\n",
    "    for(name in expt_log_pd_human_sn_demult$library){\n",
    "        print(name)\n",
    "        mol_file = file.path(base_path, name, mol_path)\n",
    "        umi_file = file.path(base_path, name, umi_path)\n",
    "        donor_ids_file = file.path(base_path, name, donor_ids_path)\n",
    "        prob_singlet_file = file.path(base_path, name, prob_singlet_path)\n",
    "        prob_doublet_file = file.path(base_path, name, prob_doublet_path)\n",
    "        \n",
    "        # read in the data and save as RDS\n",
    "        mol_data = read10xMolInfo(mol_file)\n",
    "        saveRDS(mol_data, file=paste(base_path, name, \"outs/molecule_info.rds\", sep='/'))\n",
    "\n",
    "        umi_data = Read10X_h5(umi_file)\n",
    "        saveRDS(umi_data, file=paste(base_path, name, \"outs/filtered_feature_bc_matrix.rds\", sep='/'))\n",
    "        print(\" \")\n",
    "    }\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# prob_doublet does not have the same number of columns in the data as it does in the header...\n",
    "# prob_doublet = read.table(\n",
    "#     file.path(base_path, expt_log_pd_human_sn_demult$library[[1]], prob_doublet_path), \n",
    "#     header = TRUE, \n",
    "#     sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "createMetaDataFromDCGMatrix=function(\n",
    "    expt_df,\n",
    "    umi_dcg, \n",
    "    mol_dcg,\n",
    "    donor_ids,\n",
    "    cutoff=DONOR_ASSIGNMENT_CUTOFF\n",
    "    ){\n",
    "    # takes in expt_df, a row of a dataframe \n",
    "    # umi_dcg, a dcg matrix derived from a Read 10X filtered feature matrix h5 file\n",
    "    # mol_dcg, a dcg matrix derived from a Read 10X molecule info h5 file\n",
    "    # donor_ids, a dataframe derived from vireo's donor assignment tsv file\n",
    "    # and cutoff, a float between 0 and 1, which is the cutoff for assigning a cell to a donor\n",
    "\n",
    "    # outputs a dataframe, whose rownames are the colnames of umi_dcg\n",
    "    \n",
    "    # output containing columns:\n",
    "    #   nUMI: total number of UMIs within a cell\n",
    "    #   nGene: number of unique genes expressed within a cell\n",
    "    #   nRead: number of total reads within a cell\n",
    "    #   donor_id: str, either the ID of the donor as specified in the donor_ids.tsv file, or 'unassigned'\n",
    "    #   prob_donor: float, the probability that the cell is from the donor specified in donor_id\n",
    "    #   prob_doublet: probability that the droplet represents a doublet, as determined by vireo\n",
    "    #   prob_singlet: 1-prob_doublet\n",
    "\n",
    "    #   TODO cols\n",
    "    #   dataset: str, name of dataset\n",
    "    #   orig.ident: class of data being loaded\n",
    "    #   anno_col_map: dict[str: list-of-str] for columns to be added. Keys must start with 'anno_'\n",
    "\n",
    "    umi_dcg_rownames = rownames(umi_dcg)\n",
    "    umi_dcg_colnames = colnames(umi_dcg)\n",
    "\n",
    "    # add a column to donor_ids that is the donor id as determined by the custom cutoff\n",
    "    donor_ids['donor_id_default'] = donor_ids['donor_id']\n",
    "    donor_ids['donor_id'] = 'unassigned'\n",
    "    # when the probability is above our custom cutoff, choose the best singlet\n",
    "    donor_ids$donor_id[donor_ids$prob_max >= cutoff] = (\n",
    "        donor_ids$best_singlet[donor_ids$prob_max >= cutoff])\n",
    "\n",
    "    mol_dcg_matched = mol_dcg[umi_dcg_rownames, umi_dcg_colnames]    \n",
    "    donor_ids_matched = donor_ids[umi_dcg_colnames,]\n",
    "\n",
    "    # sum within cells\n",
    "    nUMI = colSums(umi_dcg)\n",
    "    nGene = colSums(umi_dcg > 0)\n",
    "    nRead = colSums(mol_dcg_matched)\n",
    "\n",
    "    meta_df = data.frame(\n",
    "        expt_df$library,\n",
    "        expt_df$region,\n",
    "        expt_df$species,\n",
    "        expt_df$date.YYMMDD.,\n",
    "        expt_df$sub_region.tube_id,\n",
    "        expt_df$flowcell,\n",
    "        expt_df$IlluminaPlatform,\n",
    "        expt_df$Lane,\n",
    "        expt_df$num_expected_cells,\n",
    "        expt_df$transcriptome,\n",
    "        nUMI,\n",
    "        nGene, \n",
    "        nRead,\n",
    "        donor_ids_matched$donor_id, \n",
    "        donor_ids_matched$prob_max,\n",
    "        1-donor_ids_matched$prob_doublet,\n",
    "        donor_ids_matched$prob_doublet\n",
    "    )\n",
    "\n",
    "    # lowercase colnames and convert to snake_case\n",
    "    colnames(meta_df) = c(\n",
    "        \"library\", 'region', 'species', 'date', 'sub_region', 'flowcell', 'illumina_platform', \n",
    "        'lane', 'num_expected_cells', 'transcriptome',\n",
    "        \"nUMI\", \"nGene\", 'nRead', 'donor_id', 'prob_donor', 'prob_singlet', 'prob_doublet')\n",
    "    rownames(meta_df) = colnames(umi_dcg)\n",
    "    return(meta_df)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"pCalicoPDsHSrSNNURR1iPoold230214\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Setting row names on a tibble is deprecated.\"\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 2.6 GiB\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"final mol data:\"\n",
      "[1] 32734 10662\n",
      "[1] 0\n",
      "[1] 32734 10662\n",
      "[1] 0\n",
      "[1] \"Number of MT genes in the dataset: 13 / 13\"\n",
      "[1] \"Number of IEG genes in the dataset: 134 / 130\"\n",
      "[1] \"Number of OXPHOS genes in the dataset: 180 / 205\"\n",
      "[1] \"Number of rRNA genes in the dataset: 1 / 568\"\n",
      "[1] \"Number of ribosomal protein genes in the dataset: 104 / 1250\"\n",
      "[1] \"pCalicoPDsHSrSNNURR2iPoold230214\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Setting row names on a tibble is deprecated.\"\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 2.7 GiB\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"final mol data:\"\n",
      "[1] 32734 11183\n",
      "[1] 0\n",
      "[1] 32734 11183\n",
      "[1] 0\n",
      "[1] \"Number of MT genes in the dataset: 13 / 13\"\n",
      "[1] \"Number of IEG genes in the dataset: 134 / 130\"\n",
      "[1] \"Number of OXPHOS genes in the dataset: 180 / 205\"\n",
      "[1] \"Number of rRNA genes in the dataset: 1 / 568\"\n",
      "[1] \"Number of ribosomal protein genes in the dataset: 104 / 1250\"\n",
      "[1] \"pCalicoPDsHSrSNNURR3iPoold230214\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Setting row names on a tibble is deprecated.\"\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 2.9 GiB\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"final mol data:\"\n",
      "[1] 32734 11792\n",
      "[1] 0\n",
      "[1] 32734 11792\n",
      "[1] 0\n",
      "[1] \"Number of MT genes in the dataset: 13 / 13\"\n",
      "[1] \"Number of IEG genes in the dataset: 134 / 130\"\n",
      "[1] \"Number of OXPHOS genes in the dataset: 180 / 205\"\n",
      "[1] \"Number of rRNA genes in the dataset: 1 / 568\"\n",
      "[1] \"Number of ribosomal protein genes in the dataset: 104 / 1250\"\n",
      "[1] \"pCalicoPDsHSrSND9id230921D9\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Setting row names on a tibble is deprecated.\"\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 1.6 GiB\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"final mol data:\"\n",
      "[1] 32734  6426\n",
      "[1] 0\n",
      "[1] 32734  6426\n",
      "[1] 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 1.6 GiB\"\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 1.6 GiB\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Number of MT genes in the dataset: 13 / 13\"\n",
      "[1] \"Number of IEG genes in the dataset: 134 / 130\"\n",
      "[1] \"Number of OXPHOS genes in the dataset: 180 / 205\"\n",
      "[1] \"Number of rRNA genes in the dataset: 1 / 568\"\n",
      "[1] \"Number of ribosomal protein genes in the dataset: 104 / 1250\"\n",
      "[1] \"pCalicoPDsHSrSNC9id230921C9\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Setting row names on a tibble is deprecated.\"\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 1.3 GiB\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"final mol data:\"\n",
      "[1] 32734  5419\n",
      "[1] 0\n",
      "[1] 32734  5419\n",
      "[1] 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 1.3 GiB\"\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 1.3 GiB\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Number of MT genes in the dataset: 13 / 13\"\n",
      "[1] \"Number of IEG genes in the dataset: 134 / 130\"\n",
      "[1] \"Number of OXPHOS genes in the dataset: 180 / 205\"\n",
      "[1] \"Number of rRNA genes in the dataset: 1 / 568\"\n",
      "[1] \"Number of ribosomal protein genes in the dataset: 104 / 1250\"\n",
      "[1] \"pCalicoPDsHSrSNB9id230921B9\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Setting row names on a tibble is deprecated.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"final mol data:\"\n",
      "[1] 32734  2485\n",
      "[1] 0\n",
      "[1] 32734  2485\n",
      "[1] 0\n",
      "[1] \"Number of MT genes in the dataset: 13 / 13\"\n",
      "[1] \"Number of IEG genes in the dataset: 134 / 130\"\n",
      "[1] \"Number of OXPHOS genes in the dataset: 180 / 205\"\n",
      "[1] \"Number of rRNA genes in the dataset: 1 / 568\"\n",
      "[1] \"Number of ribosomal protein genes in the dataset: 104 / 1250\"\n",
      "[1] \"pCalicoPDsHSrSNA9id230921A9\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Setting row names on a tibble is deprecated.\"\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 1.9 GiB\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"final mol data:\"\n",
      "[1] 32734  7969\n",
      "[1] 0\n",
      "[1] 32734  7969\n",
      "[1] 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 1.9 GiB\"\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 1.9 GiB\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Number of MT genes in the dataset: 13 / 13\"\n",
      "[1] \"Number of IEG genes in the dataset: 134 / 130\"\n",
      "[1] \"Number of OXPHOS genes in the dataset: 180 / 205\"\n",
      "[1] \"Number of rRNA genes in the dataset: 1 / 568\"\n",
      "[1] \"Number of ribosomal protein genes in the dataset: 104 / 1250\"\n",
      "[1] \"pCalicoPDsHSrSNH8id230921H8\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Setting row names on a tibble is deprecated.\"\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 1.8 GiB\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"final mol data:\"\n",
      "[1] 32734  7226\n",
      "[1] 0\n",
      "[1] 32734  7226\n",
      "[1] 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 1.8 GiB\"\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 1.8 GiB\"\n"
     ]
    }
   ],
   "source": [
    "# for every experiment, we want to \n",
    "# 1. load the `tsv`s mapping from cells to donors\n",
    "# 2. read the molecule_info and umi data rds\n",
    "# 3. Generate col metadata\n",
    "# 4. create a singleCellExperiment object\n",
    "# 5. save the sce object as an rds file\n",
    "\n",
    "# we will be inner joining the gns to the mol_data pivoted wide using ensembl_gene_id\n",
    "# then we will left join the umi_data to the result above using gene symbol\n",
    "# TODO: resolve a number of data discrepancies\n",
    "# there are genes with symbols in the umi_data not in the EnsDb.Hsapiens.v86 \n",
    "# OR the symbols of the same genes are mismatched \n",
    "# there are genes with ensembl_gene_ids in the molecular data not in the EnsDb.Hsapiens.v86\n",
    "# there seem to be genes with duplicate symbols and different ensembl_gene_ids\n",
    "# Another discrepancy is that molecular data only includes rows with rowSums > 0, \n",
    "# whereas umi includes all irrespective of count\n",
    "\n",
    "# load human genes\n",
    "gns = as.data.frame(genes(EnsDb.Hsapiens.v86))\n",
    "gns$gene_short_name=gns$gene_name\n",
    "gns$symbol=toupper(gns$symbol)\n",
    "gns$ensembl_gene_id=row.names(gns)\n",
    "gene_names_df = gns[, c('ensembl_gene_id', 'symbol')]\n",
    "\n",
    "\n",
    "for(name in expt_log_pd_human_sn_demult$library[3:length(expt_log_pd_human_sn_demult)]){\n",
    "    print(name)\n",
    "    expt_log_row = expt_log_pd_human_sn_demult[expt_log_pd_human_sn_demult$library == name,]\n",
    "    \n",
    "    # for now, these files are not used\n",
    "    # in future iterations, we may want to add these columns to the metadata\n",
    "    #prob_singlet = read.table(file.path(base_path, name, prob_singlet_path), header = TRUE, sep = \"\\t\")\n",
    "    #prob_doublet = read.table(file.path(base_path, name, prob_doublet_path), header = TRUE, sep = \"\\t\")\n",
    "\n",
    "    # load data \n",
    "    donor_ids = read.table(file.path(base_path, name, donor_ids_path), header = TRUE, sep = \"\\t\")\n",
    "    rownames(donor_ids) = donor_ids$cell\n",
    "    mol_data = readRDS(file.path(base_path, name, 'outs/molecule_info.rds'))\n",
    "    umi_data = readRDS(file.path(base_path, name, 'outs/filtered_feature_bc_matrix.rds'))\n",
    "    # for some reason, this step duplicates some row names, which dcGMatrix allows but dataframes do not\n",
    "    rownames(umi_data) = toupper(rownames(umi_data))\n",
    "    \n",
    "    # some umi_data has repeated rownames which causes problems down the road\n",
    "    # we will sum the umi counts in these rows\n",
    "    # collect rows with duplicate rownames as a dataframe\n",
    "    n_occur = data.frame(table(rownames(umi_data)))\n",
    "    repeated_rownames = n_occur[n_occur$Freq > 1,]$Var1\n",
    "    duplicated_umi_data = as.data.frame(umi_data[rownames(umi_data) %in% repeated_rownames,])\n",
    "    duplicated_umi_data['symbol'] = sub(\"\\\\.\\\\d$\", \"\", rownames(duplicated_umi_data))\n",
    "\n",
    "    # sum across numeric columns\n",
    "    numeric_cols = setdiff(colnames(duplicated_umi_data), \"symbol\")\n",
    "    summed_duped_umi_rows = duplicated_umi_data %>%\n",
    "        group_by(symbol) %>%\n",
    "        summarise_at(vars(numeric_cols), sum, na.rm = TRUE) %>%\n",
    "        ungroup() %>%\n",
    "        as.data.frame()\n",
    "    # set rownames to be symbol column, then remove it\n",
    "    rownames(summed_duped_umi_rows) = summed_duped_umi_rows$symbol\n",
    "    summed_duped_umi_rows = summed_duped_umi_rows[, !(names(summed_duped_umi_rows) %in% c('symbol'))]\n",
    "\n",
    "    # Hack\n",
    "    # each round of the loop will remove ONE of each duplicate row \n",
    "    for(i in 1:max(n_occur$Freq)){\n",
    "        umi_data = umi_data[!rownames(umi_data) %in% repeated_rownames,]\n",
    "    }\n",
    "\n",
    "    # finally, replace the deduped rows with the summed rows via an rbind\n",
    "    umi_data = rbind(umi_data, as(summed_duped_umi_rows, \"sparseMatrix\"))\n",
    "    \n",
    "    # now we will coerce the molecular data into the size of the umi data\n",
    "    # initially this will have columns 'cell''umi''gem_group''gene''reads''library'\n",
    "    mol_data_frame = as.data.frame(mol_data[['data']])\n",
    "\n",
    "    # the $gene column are ints\n",
    "    # use the `genes` name of the mol data (which is a list of ensemble gene ids) to make that a column in the df\n",
    "    mol_data_ens_df = as.data.frame(mol_data[['genes']])\n",
    "    colnames(mol_data_ens_df) = c('ensembl_gene_id')\n",
    "    mol_data_frame['ensembl_gene_id'] = toupper(mol_data_ens_df$ensembl_gene_id[mol_data_frame$gene])\n",
    "\n",
    "    # keep only those cells that are in the umi_data\n",
    "    mol_data_frame$cell = paste0(mol_data_frame$cell, \"-1\")\n",
    "    mol_data_frame = mol_data_frame[mol_data_frame$cell %in% colnames(umi_data), ]\n",
    "\n",
    "    # Group by 'cell' and 'ensembl_gene_id' and sum the number of occurrences\n",
    "    grouped_df = mol_data_frame %>%\n",
    "        group_by(cell, ensembl_gene_id) %>%\n",
    "        summarize(reads = sum(reads), .groups = \"drop\")\n",
    "\n",
    "    mol_data_wide = grouped_df %>% \n",
    "        pivot_wider(names_from = cell, values_from = reads, values_fill = 0)\n",
    "\n",
    "    # we need the gene names for these ensemble gene ids\n",
    "    # NOTE: IT SEEMS THERE ARE SOME ENSEMBLE GENE IDS THAT ARE NOT IN THE GENE NAMES DATAFRAME\n",
    "    mol_data_wide_joined = merge(mol_data_wide, gene_names_df, by='ensembl_gene_id')\n",
    "\n",
    "    # find genes with multiple ensemble_gene_ids per symbol\n",
    "    duplicate_gene_symbols = mol_data_wide_joined %>% \n",
    "        group_by(symbol) %>% \n",
    "        summarize(n=n(), .groups = \"drop\") %>% \n",
    "        filter(n > 1) %>% \n",
    "        arrange(desc(n))\n",
    "\n",
    "    # isolate the rows corresponding to these duplicated symbols and sum across them\n",
    "    dupe_rows = mol_data_wide_joined[mol_data_wide_joined$symbol %in% duplicate_gene_symbols$symbol,]\n",
    "    sum_dupe_rows = dupe_rows %>%\n",
    "        group_by(symbol) %>%\n",
    "        summarize(across(-c('ensembl_gene_id'), sum))\n",
    "    rownames(sum_dupe_rows) = sum_dupe_rows$symbol\n",
    "\n",
    "    # remove rows with duplicate gene symbols\n",
    "    mol_data_wide_joined_deduped = mol_data_wide_joined[!mol_data_wide_joined$symbol %in% duplicate_gene_symbols$symbol,]\n",
    "    # drop the ensembl_gene_id column, which is now redundant\n",
    "    mol_data_wide_joined_deduped = mol_data_wide_joined_deduped[, !(names(mol_data_wide_joined_deduped) %in% c('ensembl_gene_id'))]\n",
    "    # now symbol is unambiguous, so we can use it as the rownames\n",
    "    rownames(mol_data_wide_joined_deduped) = mol_data_wide_joined_deduped$symbol \n",
    "\n",
    "    # add back in the summed reads to the mol_data_wide_joined \n",
    "    mol_data_wide_joined_deduped_add_back = rbind(mol_data_wide_joined_deduped, sum_dupe_rows)\n",
    "\n",
    "    # type coercing to data.frame\n",
    "    umi_df = as.data.frame(umi_data)\n",
    "    umi_df['symbol'] = rownames(umi_df)\n",
    "    umi_df = as.data.frame(umi_df[, c('symbol')])\n",
    "    colnames(umi_df) = c('symbol')\n",
    "\n",
    "    # finally we want to left join the umi_data to the result above using gene symbol\n",
    "    mol_data_final = left_join(umi_df, mol_data_wide_joined_deduped_add_back, by='symbol')\n",
    "    rownames(mol_data_final) = mol_data_final$symbol\n",
    "    mol_data_final = mol_data_final[, !(names(mol_data_final) %in% c('symbol'))]\n",
    "\n",
    "    # convert all nans to 0s\n",
    "    mol_data_final[is.na(mol_data_final)] = 0\n",
    "\n",
    "\n",
    "    meta_df = createMetaDataFromDCGMatrix(\n",
    "        expt_log_row,\n",
    "        umi_data, \n",
    "        mol_data_final,\n",
    "        donor_ids,\n",
    "        cutoff=DONOR_ASSIGNMENT_CUTOFF)\n",
    "\n",
    "    data_sce=.myExpSetCreatorFn(inputExpData=umi_data,\n",
    "                            organism=\"human\",\n",
    "                            minExpCells=0,\n",
    "                            inputPdata=meta_df,\n",
    "                            inputFdata=NULL,\n",
    "                            addExtraAnno=T,\n",
    "                            ncores=12)\n",
    "    saveRDS(data_sce, paste0(\"~/seq_data/pd/\", name, \"/data_sce.rds\"))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "sce_list = list()\n",
    "for(name in expt_log_pd_human_sn_demult$library){\n",
    "    print(name)\n",
    "    fname = file.path(\"~/seq_data/pd\", name, \"/data_sce.rds\")\n",
    "    sce = readRDS(fname)\n",
    "    sce_list = append(sce_list, list(sce))\n",
    "}\n",
    "pilot_sce = .mycBindFn(sce_list)\n",
    "dim(pilot_sce)\n",
    "saveRDS(pilot_sce, paste0(\"~/seq_data/pd/pilot_sce.rds\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"wide data, joined to genes, summed by symbol, 0-rows added back:\"\n",
      "[1] 31754 12991\n",
      "[1] \"UMI data:\"\n",
      "[1] 32738 12990\n"
     ]
    }
   ],
   "source": [
    "# molecular data only includes rows with rowSums > 0\n",
    "# umi data has all rows, including for genes that were not expressed in the library\n",
    "# add back in those rows with rowsums == 0 to the mol_data_wide_joined_deduped\n",
    "umi_rowsums_zero_rownames = toupper(rownames(umi_data))[rowSums(umi_data) == 0]\n",
    "\n",
    "# Create a matrix of zeros\n",
    "zero_matrix = matrix(0, nrow=length(umi_rowsums_zero_rownames), ncol=length(colnames(mol_data_wide_joined_deduped_add_back)),\n",
    "                      dimnames=list(umi_rowsums_zero_rownames, colnames(mol_data_wide_joined_deduped_add_back)))\n",
    "# Convert matrix to dataframe\n",
    "rowsums_zero_df = as.data.frame(zero_matrix)\n",
    "\n",
    "mol_data_wide_joined_deduped_w_zeros = rbind(mol_data_wide_joined_deduped_add_back, rowsums_zero_df)\n",
    "\n",
    "print(\"wide data, joined to genes, summed by symbol, 0-rows added back:\")\n",
    "print(dim(mol_data_wide_joined_deduped_w_zeros))\n",
    "print(\"UMI data:\")\n",
    "print(dim(umi_data))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
